Add Hyperion config flow and tests

dermotduffy opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
Adds Hyperion config flow, with automated migration between YAML configuration to config entries. Every attempt is made to preserve entity_ids, but they may change during this migration.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
   * Add a Hyperion configuration flow, with tests. This is part 2 of the major rework of the Hyperion component (see [Part 1](https://github.com/home-assistant/core/pull/39738)).
   * This will allow the YAML configuration to be completely removed in a subsequent PR (after the deprecation window).
   * Rework tests to use config entry based testing by default.
   * This automatically pulls in new features that the underlying library supports but that were not available before due to limitations of the YAML configuration support (e.g. access to multiple Hyperion instances on the same server, Hyperion authentication, dynamic instances -- entities created/removed in Home Assistant as they are created/removed on the Hyperion server).
   * Add an option (priority) that can be tweaked from the UI.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [X] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:

Good news: This section now intentionally blank!

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 39738
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/15336

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

dermotduffy committed:
Add initial skeletal config flow.

dermotduffy committed:
Deprecate all remaining YAML options.

dermotduffy committed:
Add basic config_entry plumbing.

dermotduffy committed:
Add priority and name plumbing.

dermotduffy committed:
Add token support.

dermotduffy committed:
Add auth handling.

dermotduffy committed:
Test correct abort message is used.

dermotduffy committed:
Variable tidy.

dermotduffy committed:
Break out instance into a separate stage/form.

dermotduffy committed:
Add zeroconf step.

dermotduffy committed:
Add flow diagram.

dermotduffy committed:
Use IP address rather than hostname for zeroconf.

dermotduffy committed:
Add zeroconf support and final confirmation.

dermotduffy committed:
Update http port comment and remove TODO for now.

dermotduffy committed:
Back out instances support, the entry is for the Hyperion server.

dermotduffy committed:
Dynamic instance creation/deletion.

dermotduffy committed:
Convert tests to use config entries by default

dermotduffy committed:
Test unload.

dermotduffy committed:
Argument cleanup.

dermotduffy committed:
Test YAML key

dermotduffy committed:
Add automatic migration from YAML to config entry

dermotduffy committed:
Various pylint fixes

dermotduffy committed:
Remove completed TODO

dermotduffy committed:
Add unique_id comment.

dermotduffy committed:
Add to the domain to the id for extensibility.

dermotduffy committed:
Update unique_id comment.

dermotduffy committed:
Add skeletal options flow.

dermotduffy committed:
Add strings for options flow.

dermotduffy committed:
Plumb options through to entity.

dermotduffy committed:
Add strings entry for already_in_progress.

dermotduffy committed:
Set unique_id as early as possible for Zeroconf.

dermotduffy committed:
Ensure entry_id actually exists in data.

dermotduffy committed:
Fix bug in dynamic token creation.

dermotduffy committed:
Use context manager for connections.

dermotduffy committed:
Lint fixes

dermotduffy committed:
Migrate options from YAML.

dermotduffy committed:
Add SSDP support.

dermotduffy committed:
Fix duplicate SSDP discoveries.,

dermotduffy committed:
Remove zeroconf support (as SSDP is used anyway).

dermotduffy committed:
Fix migration bug of old unique_id values.

dermotduffy committed:
Add test for new style unique_id without config

dermotduffy committed:
Remove domain from unique_id.

dermotduffy committed:
Add log entry for migration

dermotduffy committed:
Change title to host:port rather than id.

dermotduffy committed:
Rely on the latest hyperion-py

dermotduffy committed:
Fix merge error AsyncMock -> CoroutineMock.

dermotduffy commented:
If @MartinHjelmare had the time, they may be a great person to review this change given the context they have and the great review they did for the earlier [Hyperion Refactor Part 1](https://github.com/home-assistant/core/pull/39738). No worries if not!

dermotduffy committed:
Test some additional error conditions.

balloob commented on homeassistant/components/hyperion/__init__.py:
```diff
@@ -1 +1,72 @@
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Hyperion from a config entry."""
+    for component in PLATFORMS:
```
We should test here if the connection can be established and raise `ConfigEntryNotReady` when it can't be established.

balloob replied:
And if we detect that the auth is no longer valid, we should `return False` instead of raising. In a future PR you can then add a `reauth` step.

dermotduffy replied:
Done (Hyperion auth does not expire, although can be manually revoked -- might be something to tackle later).

balloob commented on homeassistant/components/hyperion/config_flow.py:
```diff
@@ -0,0 +1,422 @@
+#           |                   |
+#           v                   |
+#     +----------------+        |
+#     | Step: Confirm  |<-------+
```
1. User shouldn't need to confirm after already entering their host
2. The SSDP step should directly jump to auth or confirm step

balloob commented on homeassistant/components/hyperion/config_flow.py:
```diff
@@ -0,0 +1,422 @@
+
+        if auth_resp.get(const.KEY_INFO, {}).get(const.KEY_REQUIRED) is True:
+            # Auth is required, show the form to get the token.
+            return self._show_auth_form()
+        return await self._show_confirm_form()
```
Don't directly call the auth form function but instead delegate that to the auth step.

```suggestion
            return await self.async_step_auth()
        return await self.async_step_confirm()
```

balloob commented on homeassistant/components/hyperion/config_flow.py:
```diff
@@ -0,0 +1,422 @@
+        return HyperionOptionsFlow(config_entry)
+
+
+class HyperionOptionsFlow(config_entries.OptionsFlow):
```
Please keep this PR simpler and keep options flow for a future PR

dermotduffy replied:
If the options are excluded and saved for a future PR, anyone who migrates from YAML -> ConfigEntry in the intervening time will lose the value of the priority option (which is currently in the YAML).

With that said -- please let me know if you'd rather I proceed with splitting this out and I will do so.

MartinHjelmare replied:
We can use a different target branch than the dev branch and do multiple PRs to this other target branch. Then when ready for release we make a PR from the other target branch to dev branch.

Let me know and I can push a copy of dev as a new branch and you can rebase and change target branch here.

balloob replied:
It could still be imported, just not changed via an options flow.

dermotduffy replied:
I think the choice is complexity in the codereview (multiple branches) vs complexity in the code (an extra scenario to consider, where the config is migrated but the options not).

Lets go with the new branch if that is okay with you both? @MartinHjelmare

balloob commented on homeassistant/components/hyperion/light.py:
```diff
@@ -78,27 +101,239 @@
+    # The unique_id needs to be updated, then the config_flow should do the rest.
+    registry = await async_get_registry(hass)
+    for entity_id, entity in registry.entities.items():
+        if entity.config_entry_id is None and entity.platform == DOMAIN:
```
Use a guard clause to make it slightly easier to see what's going on
```python
        if entity.config_entry_id is not None or entity.platform != DOMAIN:
            continue
```

balloob commented on homeassistant/components/hyperion/light.py:
```diff
@@ -78,27 +101,239 @@
+            desired_instance_ids.add(instance_id)
+            unique_id = get_hyperion_unique_id(server_id, instance_id)
+            entity_id = registry.async_get_entity_id(LIGHT_DOMAIN, DOMAIN, unique_id)
+            if entity_id is not None and entity_id in platform.entities:
```
This check is not good. Users can disable entities in the entity registry and then they will not end up in `platform.entities`. If you want to prevent from adding something twice, instead of relying on HA data, keep a set around to track `(server_id, instance_id)` tuples.

balloob commented on homeassistant/components/hyperion/light.py:
```diff
@@ -78,27 +101,239 @@
+            )
+
+        # Delete instances that are no longer present on this server.
+        hyperion_entity_ids = list(platform.entities.keys())
```
This won't work because of disabling in entity registry too. Also needs to avoid using the entity platform.

balloob commented on homeassistant/components/hyperion/light.py:
```diff
@@ -78,27 +101,239 @@
+        async_add_entities(entities_to_add)
+
+    hyperion_client = await _async_create_connect_client(host, port, token=token)
+    if not hyperion_client:
+        raise PlatformNotReady
+    server_id = await hyperion_client.async_id()
+    if not server_id:
+        await hyperion_client.async_client_disconnect()
         raise PlatformNotReady
+    hass.data[DOMAIN][config_entry.entry_id] = hyperion_client
```
Move this logic to `__init__.py` `async_setup_entry`

balloob commented on homeassistant/components/hyperion/light.py:
```diff
@@ -78,27 +101,239 @@
+    hyperion_client = client.HyperionClient(
+        host,
+        port,
+        **{CONF_TOKEN: token, CONF_INSTANCE: instance},
```
Why not write this?

```suggestion
        token=token, instance=instance,
```

balloob commented on homeassistant/components/hyperion/light.py:
```diff
@@ -78,27 +101,239 @@
+    return hyperion_client
+
+
+async def async_unload_entry(hass: HomeAssistantType, entry: ConfigEntry) -> bool:
```
This should be moved to `__init__`

balloob commented on homeassistant/components/hyperion/light.py:
```diff
@@ -375,3 +615,7 @@ async def async_added_to_hass(self):
+
+    async def async_will_remove_from_hass(self):
+        """Disconnect from server."""
+        await self._client.async_client_disconnect()
```
Isn't the client shared across all entities ?

dermotduffy replied:
No. Each entity has its own client. A Hyperion server supports multiple instances. The client connects to the server -- then selects an instance, and receives subscriptions, in the context of that instance. As an instance == an entity, there is thus one client per entity.

balloob commented on homeassistant/components/hyperion/strings.json:
```diff
@@ -0,0 +1,55 @@
+{
+  "config": {
+    "flow_title": "Hyperion Ambilight",
```
No need to add this if it is the same as the manifest title.

balloob commented on homeassistant/components/hyperion/strings.json:
```diff
@@ -0,0 +1,55 @@
+    "flow_title": "Hyperion Ambilight",
+    "step": {
+      "user": {
+        "description": "Set up your Hyperion Ambilight with Home Assistant",
```
We can drop this too. This is implied.

balloob commented on homeassistant/components/hyperion/strings.json:
```diff
@@ -0,0 +1,55 @@
+        }
+      },
+      "create_token": {
+        "description": "Choose **Submit** below to request a new authentication token. You will be redirected to the [Hyperion UI]({hyperion_url}) to approve the request. Please verify the shown id is \"{auth_id}\"",
```
Seems confusing to make this a link since there is already an external step button that they should press.

balloob commented on homeassistant/components/hyperion/strings.json:
```diff
@@ -0,0 +1,55 @@
+      }
+    },
+    "error": {
+      "connection_error": "Failed to connect to Hyperion Ambilight device",
```
Use references here.

balloob commented on homeassistant/components/hyperion/strings.json:
```diff
@@ -0,0 +1,55 @@
+      "create_token_error": "Failed to create new token"
+    },
+    "abort": {
+      "already_configured": "The Hyperion Ambilight server is already configured",
```
Same

dermotduffy committed:
Phase 1 fixes for codereview

dermotduffy committed:
Rearrange flow steps as per code review.

dermotduffy committed:
Update flow diagram

dermotduffy committed:
Minor diagram update

dermotduffy committed:
Re-arrange loading logic  as per codereview.

dermotduffy committed:
Remove TODO.

dermotduffy committed:
Remove options flow and save for a future PR.

dermotduffy committed:
Remove TODO as it fails pylint.

dermotduffy commented:
Thanks for the great review @balloob ! In particular, thanks for the guidance of avoiding use of the entity_platform (now not directly used at all). I believe I have addressed all your comments, and that this is ready for another look.

Once you're happy, if yourself or @MartinHjelmare making a temporary branch for me, I'll rebase onto it and adjust the PR target. Once that's in, I'll add another PR to restore the (now removed) options into that same temporary branch.

Thanks again!

MartinHjelmare commented:
I've pushed a fresh copy of dev branch here:
https://github.com/home-assistant/core/tree/add-hyperion-config-options-flow

dermotduffy commented:
Thanks @MartinHjelmare ! I have now rebased onto that branch, force pushed and updated the PR target here.

From my perspective, this is now ready to merge into `add-hyperion-config-options-flow`, if you are both happy with the modifications I've made to address the code review feedback.

balloob commented on homeassistant/components/hyperion/config_flow.py:
```diff
@@ -0,0 +1,379 @@
+
+    async def _create_client(
+        self, raw_connection=False
+    ) -> Optional[client.HyperionClient]:
```
```suggestion
    ) -> client.HyperionClient:
```

balloob commented on homeassistant/components/hyperion/light.py:
```diff
@@ -78,27 +106,190 @@
+    config_entry.add_update_listener(_async_options_updated)
+
+
+async def async_unload_entry(hass: HomeAssistantType, config_entry: ConfigType):
```
This doesn't exist.

balloob reviewed (approved):
Looks great. 2 minor comments. 🎉

MartinHjelmare commented on tests/components/hyperion/test_light.py:
```diff
@@ -1,82 +1,278 @@
+
+    # There should be a config entry with the correct server unique_id.
+    entry_id = next(iter(hass.data[DOMAIN]))
+    assert client == hass.data[DOMAIN][entry_id][CONF_ROOT_CLIENT]
```
Please don't assert any integration details.

https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations

MartinHjelmare commented on tests/components/hyperion/test_light.py:
```diff
@@ -420,11 +622,26 @@ async def test_full_state_loaded_on_start(hass):
+    assert client.async_client_connect.called
+    assert not client.async_client_disconnect.called
+
+    assert await async_unload_entry(hass, entry)
```
Only use the core interfaces to start the tests. Here use `hass.config_entries.async_unload`.

MartinHjelmare commented on tests/components/hyperion/test_light.py:
```diff
@@ -420,11 +622,26 @@ async def test_full_state_loaded_on_start(hass):
+
+    entry = await setup_test_config_entry(hass)
+    assert hass.states.get(TEST_ENTITY_ID_1) is not None
+    client = hass.data[DOMAIN][entry.entry_id][CONF_ROOT_CLIENT]
+    assert client.async_client_connect.called
+    assert not client.async_client_disconnect.called
```
Patch the library client instead and assert the calls to the mock. Don't assert integration details.

MartinHjelmare commented on homeassistant/components/hyperion/config_flow.py:
```diff
@@ -0,0 +1,379 @@
+_LOGGER = logging.getLogger(__name__)
+_LOGGER.setLevel(logging.DEBUG)
+
+#  +------------------+    +------------------+    +--------------------+
```
Will this comment really be kept updated as the integration evolves?

dermotduffy replied:
I appreciate this question. 

I think the diagram makes it easier for people not familiar with the flow to understand what's happening. The Hyperion component has barely changed in the past years (except large scale codebase changes) until I became the only code owner a few weeks back.

In short: I think it's better with the diagram. The codebase rarely ever changes. I think it's worth trying to keep it up to date!

MartinHjelmare commented on homeassistant/components/hyperion/config_flow.py:
```diff
@@ -0,0 +1,379 @@
+
+    def __init__(self) -> None:
+        """Instantiate config flow."""
+        super().__init__()
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/hyperion/config_flow.py:
```diff
@@ -0,0 +1,379 @@
+        # authentication is required.
+        async with await self._create_client(raw_connection=True) as hyperion_client:
+            if not hyperion_client:
+                return self._show_setup_form({CONF_BASE: "connection_error"})
```
We can't reuse the form for the user step in other steps. The form is step specific and we should only move between step functions by await a step function, ie not go directly to another steps' form.

MartinHjelmare commented on homeassistant/components/hyperion/config_flow.py:
```diff
@@ -0,0 +1,379 @@
+            title=f"{self._data[CONF_HOST]}:{self._data[CONF_PORT]}", data=self._data
+        )
+
+    def _show_setup_form(self, errors: Optional[Dict] = None) -> Dict[str, Any]:
```
There's rarely a reason to break out the form calls into methods since they are not reusable by other steps. I suggest we inline these in the step methods where they belong.

dermotduffy committed:
Address @balloob code review feedback

dermotduffy commented on homeassistant/components/hyperion/__init__.py:
```diff
@@ -1 +1,132 @@
+# Unique ID
+# =========
+# A config entry represents a connection to a single Hyperion server. The config
+# entry_id is the server id returned from the Hyperion instance (a unique ID per
```
Typo!

dermotduffy commented on tests/components/hyperion/test_config_flow.py:
```diff
@@ -0,0 +1,522 @@
+    result = await _init_flow(hass)
+
+    client = create_mock_client()
+    with patch("hyperion.client.HyperionClient", return_value=client):
```
Done (in the common _configure_flow function).

dermotduffy committed:
Codereview fixes for @MartinHjelmare (Part 1)

dermotduffy committed:
Codereview fixes for @MartinHjelmare (Part 2)

dermotduffy committed:
Codereview fixes for @MartinHjelmare (Part 3)

dermotduffy committed:
pylint fix.

dermotduffy commented:
Thank you @balloob and @MartinHjelmare ! Fixes pushed and ready for another look.

MartinHjelmare commented on homeassistant/components/hyperion/light.py:
```diff
@@ -362,6 +559,13 @@ def _update_client(self, json):
     async def async_added_to_hass(self):
         """Register callbacks when entity added to hass."""
+
+        self._dispatcher_instance_removed_unsub = async_dispatcher_connect(
+            self.hass,
+            SIGNAL_INSTANCE_REMOVED.format(self._unique_id),
+            self.async_remove,
+        )
```
We can use `self.async_on_remove` to register the unsubscribe callbacks for removal on entity remove.
```suggestion
        self.async_on_remove(
            async_dispatcher_connect(
                self.hass,
                SIGNAL_INSTANCE_REMOVED.format(self._unique_id),
                self.async_remove,
            )
        )
```

MartinHjelmare commented on homeassistant/components/hyperion/light.py:
```diff
@@ -375,3 +579,11 @@ async def async_added_to_hass(self):
+    async def async_will_remove_from_hass(self):
+        """Disconnect from server."""
+        await self._client.async_client_disconnect()
+
+        if self._dispatcher_instance_removed_unsub:
+            self._dispatcher_instance_removed_unsub()
+            self._dispatcher_instance_removed_unsub = None
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/hyperion/light.py:
```diff
@@ -78,34 +110,194 @@
             brightness=255, rgb_color=DEFAULT_COLOR, effect=KEY_EFFECT_SOLID
         )
         self._effect_list = []
+        self._dispatcher_instance_removed_unsub = None
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/hyperion/light.py:
```diff
@@ -78,34 +110,194 @@
+        registry = await async_get_registry(hass)
+        entities_to_add: List[Hyperion] = []
+        desired_unique_ids: Set[str] = set()
+        current_entities = hass.data[DOMAIN][config_entry.entry_id][LIGHT_DOMAIN][
```
We don't need to define this variable again here. We can access the variable that is available in the outer scope.

dermotduffy replied:
@MartinHjelmare -- can you clarify, which variable are you referring to? If it's `current_entities`, then this is really just a convenience to not need to use hass.data[...][...][...][...], and I don't see this readily available at the outer scope?

dermotduffy replied:
Ah, I think you're referring to `current_entities` which _could be_ at the outer scope, as per your suggestion below.

MartinHjelmare commented on homeassistant/components/hyperion/light.py:
```diff
@@ -78,34 +110,194 @@
+
+        async_add_entities(entities_to_add)
+
+    hass.data[DOMAIN][config_entry.entry_id][LIGHT_DOMAIN] = {CONF_ENTITIES: set()}
```
I don't think we even need to store the container in hass.data. The variable should be kept around in memory since it's accessed by a function that we register and store.
```suggestion
    current_entities = set()
```

MartinHjelmare commented on tests/components/hyperion/test_light.py:
```diff
@@ -420,11 +633,25 @@ async def test_full_state_loaded_on_start(hass):
+    entry = _get_config_entry_from_unique_id(hass, TEST_SERVER_ID)
+    assert entry
+
+    await entry.async_unload(hass)
```
```suggestion
    await hass.config_entries.async_unload(entry.entry_id)
```

MartinHjelmare commented on homeassistant/components/hyperion/config_flow.py:
```diff
@@ -0,0 +1,380 @@
+    async def async_step_user(
+        self,
+        user_input: Optional[ConfigType] = None,
+        errors: Optional[Dict] = None,
```
The step functions should not accept other parameters than the user_input and the form errors are private to each step since they are specific to the form.

dermotduffy replied:
@MartinHjelmare -- help me out here, as it feels like these constraints:

   * An inability to pass in errors into a step.
   * An inability to return another step's form.

... make it impossible (?) to show an error message to a user in the context of  a different step.

An example:
   * The `user` steps get host/port (may or may not be verified at the time).
   * A future step uses that host/port, but fails to connect.

Desired outcome:
   * We'd want  to return the user to the host/port choice, with an error message saying `Can't connect`.

How can I achieve this, or do you have an alternative suggestion?

MartinHjelmare replied:
If the user step form fails there should not be a different step after the user step. Each step should be validated before moving on to the next step. We should never need to return to a previous step.

dermotduffy replied:
OK. What behavior would you like if the subsequent connection fails in a future step, even though it was verified in a prior step? Abort?

MartinHjelmare replied:
Yeah, since that doesn't have to do with the user input.

MartinHjelmare commented on homeassistant/components/hyperion/config_flow.py:
```diff
@@ -0,0 +1,380 @@
+        # used to open a URL, that the user already knows the address of).
+        return f"http://{self._data[CONF_HOST]}:{self._port_ui}"
+
+    async def async_step_auth(
```
I'd suggest to make this step a helper method without a form instead. It can then return different error reasons that the calling step can turn into form errors and show.

dermotduffy replied:
But this step has a form it needs (static token vs create a token) -- I don't follow how that can be removed.

Perhaps you mean pull the top half of this step (where we see if auth is actually required) into a helper method, and that decides whether or not to invoke the auth step in the first place to gather the required data in the auth form?

MartinHjelmare replied:
Yeah, isolate the action that decides the next step from the form steps.

dermotduffy committed:
Codereview fixes for @MartinHjelmare (Part 4)

dermotduffy commented:
@MartinHjelmare -- easy items done, request for clarification on how you'd like config_flow changes to be implemented.

dermotduffy committed:
Codereview fixes for @MartinHjelmare (Part 4)

dermotduffy committed:
pylint fixes.

dermotduffy commented:
@MartinHjelmare This is ready for another look. The main change is a config_flow.py update so the steps are more self-contained (i.e. never refer to other steps favoring abort rather than errors  of another step) and more use of helper methods that do not involve a form. Thanks for your patience in this review.

MartinHjelmare commented on homeassistant/components/hyperion/config_flow.py:
```diff
@@ -0,0 +1,395 @@
+
+            # Cancellation is only processed on the **next** cycle of the event loop, so
+            # yield here.
+            await asyncio.sleep(0)
```
Isn't it enough to await the task? That also yields to the loop.

MartinHjelmare commented on homeassistant/components/hyperion/config_flow.py:
```diff
@@ -0,0 +1,395 @@
+            await self.hass.config_entries.flow.async_configure(
+                flow_id=self.flow_id, user_input=auth_resp
+            )
+        return auth_resp
```
Nothing is checking this return value.

MartinHjelmare commented on homeassistant/components/hyperion/config_flow.py:
```diff
@@ -0,0 +1,395 @@
+        self._require_confirm = False
+        self._port_ui = const.DEFAULT_PORT_UI
+
+    async def _create_client(self, raw_connection=False) -> client.HyperionClient:
```
This doesn't need to be a coroutine function since we don't await inside.

MartinHjelmare commented on homeassistant/components/hyperion/config_flow.py:
```diff
@@ -0,0 +1,395 @@
+            return await self._advance_to_auth_step_if_necessary(hyperion_client)
+
+    async def async_step_ssdp(
+        self, discovery_info: Optional[ConfigType] = None
```
Discovery info is not optional.

MartinHjelmare commented on homeassistant/components/hyperion/config_flow.py:
```diff
@@ -0,0 +1,395 @@
+            self._data[CONF_PORT] = const.DEFAULT_PORT_JSON
+
+        hyperion_id = discovery_info.get(ATTR_UPNP_SERIAL)
+        if hyperion_id:
```
Invert this check and abort if true. Then we can outdent below.

MartinHjelmare commented on homeassistant/components/hyperion/config_flow.py:
```diff
@@ -0,0 +1,395 @@
+            if user_input.get(CONF_CREATE_TOKEN):
+                self._auth_id = client.generate_random_auth_id()
+                return self.async_show_form(
+                    step_id="create_token",
```
Don't move directly to another step's form. Await the step function.

MartinHjelmare commented on homeassistant/components/hyperion/config_flow.py:
```diff
@@ -0,0 +1,395 @@
+        if not hyperion_id:
+            return self.async_abort(reason="no_id")
+
+        await self.async_set_unique_id(hyperion_id)
```
We may want to set `raise_on_progress` parameter to False here if the user starts a flow and there's already a discovery flow in progress for the same device.

dermotduffy replied:
Good catch.

MartinHjelmare commented on homeassistant/components/hyperion/strings.json:
```diff
@@ -0,0 +1,53 @@
+      }
+    },
+    "error": {
+      "connection_error": "[%key:common::config_flow::error::cannot_connect%]",
```
Please use our standard error key names, eg `cannot_connect`.

dermotduffy committed:
Codereview fixes for @MartinHjelmare (Part 5)

MartinHjelmare commented on homeassistant/components/hyperion/config_flow.py:
```diff
@@ -121,13 +121,13 @@ async def async_step_import(
 
     async def async_step_ssdp(
-        self, discovery_info: Optional[ConfigType] = None
+        self, discovery_info: ConfigType = None
```
```suggestion
        self, discovery_info: ConfigType
```

dermotduffy commented:
Thanks @MartinHjelmare ! Part 5 fixes committed -- ready for a look.

dermotduffy committed:
Codereview fixes for @MartinHjelmare (Part 6)

MartinHjelmare reviewed (approved):
Great!

dermotduffy commented:
(The 'Close' button is very large on the github mobile app, fat finger!)

Thank you very much @MartinHjelmare for the high-quality review. The code has significantly improved as a result of your and @balloob 's reviews, and I've learned a lot about how HA works and how components and config_flows should be written.

I'm not quite sure what the protocol here is given the custom branch (`home-assistant:add-hyperion-config-options-flow`). I am assuming you'll merge it in once you're ready, and I should prepare the options PR against that same branch?

MartinHjelmare commented:
I've rebased the target branch on latest dev branch.

dermotduffy commented:
@MartinHjelmare Thank you! Build passed, ready to go!
